File: web/app/src/api/meta/captcha/siteKeys.ts
import v1 from "../../v1";
export class CaptchaSiteKeys {
  siteKeys: {
    easy: string;
    moderate: string;
    hard: string;
    all: string;
  };
  difficulties: {
    user: {
      create: string;
      emailToken: string;
    };
  };

  static createFrom(source: Partial<CaptchaSiteKeys> = {}) {
    return new CaptchaSiteKeys(source);
  }

  constructor(source: Partial<CaptchaSiteKeys> = {}) {
    if (typeof source === "string") source = JSON.parse(source);
    this.siteKeys = source["site-keys"];
    this.difficulties = source["difficulties"];
  }

  siteKey(key: "easy" | "moderate" | "hard" | "all" | string): string {
    let siteKey = this.siteKeys.all;
    if (this.siteKeys[key] != undefined) {
      siteKey = this.siteKeys[key];
    }
    return siteKey;
  }
}

export async function siteKeys(): Promise<CaptchaSiteKeys> {
  return v1.get("meta/captcha/site-keys").then((r) => {
    return CaptchaSiteKeys.createFrom(r.data);
  });
}

File: internal/router/handlers/api/v1/meta/captcha/siteKeys.go
package captcha

import (
	"github.com/allocamelus/allocamelus/internal/g"
	"github.com/allocamelus/allocamelus/pkg/fiberutil"
	"github.com/gofiber/fiber/v2"
)

type siteKeysResp struct {
	SiteKeys     siteKeys               `json:"site-keys"`
	Difficulties map[string]interface{} `json:"difficulties"`
}

type siteKeys struct {
	Easy     string `json:"easy"`
	Moderate string `json:"moderate"`
	Hard     string `json:"hard"`
	All      string `json:"all"`
}

// SiteKeys site key handler
func SiteKeys(c *fiber.Ctx) error {
	return fiberutil.JSON(c, 200, siteKeysResp{
		SiteKeys: siteKeys{
			Easy:     g.Config.HCaptcha.Easy,
			Moderate: g.Config.HCaptcha.Moderate,
			Hard:     g.Config.HCaptcha.Hard,
			All:      g.Config.HCaptcha.All,
		},
		Difficulties: map[string]interface{}{
			"user": map[string]string{
				"create":     "moderate",
				"emailToken": "moderate",
			},
		},
	})
}
